src/text/writer.rs

```diff
@@ -0,0 +1,352 @@
+use crate::result::{illegal_operation, IonResult};
+use crate::text::encoder::TextEncoder;
+use crate::IonType;
+use bigdecimal::BigDecimal;
+use chrono::{DateTime, FixedOffset};
+use std::io::{BufWriter, Write};
+
+pub struct TextWriter<W: Write> {
+    output: BufWriter<W>,
+    annotations: Vec<String>,
+    field_name: Option<String>,
+    containers: Vec<IonType>,
+}
+
+impl<W: Write> TextWriter<W> {
+    /// Constructs a new instance of TextWriter that writes values to the provided io::Write
+    /// implementation.
+    pub fn new(sink: W) -> TextWriter<W> {
+        TextWriter {
+            output: BufWriter::new(sink),
+            annotations: vec![],
+            field_name: None,
+            containers: vec![],
+        }
+    }
+
+    /// Returns a reference to the underlying io::Write implementation.
+    pub fn output(&self) -> &W {
+        self.output.get_ref()
+    }
+
+    /// Causes any buffered data to be written to the underlying io::Write implementation.
+    pub fn flush(&mut self) -> IonResult<()> {
+        self.output.flush()?;
+        Ok(())
+    }
+
+    /// Sets the current field name to `name`. If the TextWriter is currently positioned inside
+    /// of a struct, the field name will be written before the next value. Otherwise, it will be
+    /// ignored.
+    pub fn set_field_name(&mut self, name: &str) {
+        self.field_name = Some(name.to_string());
+    }
+
+    /// Sets a list of annotations that will be applied to the next value that is written.
+    pub fn set_annotations(&mut self, annotations: &[&str]) {
+        self.annotations
+            .extend(annotations.iter().map(|s| s.to_string()));
+    }
```

Comment:
As you know, it would be nice if we didn't have to do the copy here :wink:

Reply:
Yep. I could eagerly write out the `field_name` and `annotations` as soon as the user provides them, but I'd need to rewrite things to track which function calls would be legal in which states. I also considered keeping a small pool of reusable annotations `String`s to avoid the cost of new allocations. Both seemed like more effort than they were worth for an initial pass at the less-performance-sensitive text writer. I'll open an issue to track it.

Reply:
See issue #92.